File: gto-support-db/src/test/java/org/ccci/gto/android/common/db/ExpressionTest.java
package org.ccci.gto.android.common.db;

import android.text.TextUtils;
import android.util.Pair;

import org.ccci.gto.android.common.db.model.Root;
import org.ccci.gto.android.common.testing.CommonMocks;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import static org.ccci.gto.android.common.db.Contract.RootTable;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(PowerMockRunner.class)
@PrepareForTest({Pair.class, TextUtils.class})
public class ExpressionTest {
    private AbstractDao mDao;

    @Before
    public void setup() throws Exception {
        CommonMocks.mockPair();
        CommonMocks.mockTextUtils();

        mDao = mock(AbstractDao.class);
        when(mDao.getTable(eq(Root.class))).thenReturn(RootTable.TABLE_NAME);
    }

    @Test
    public void testEqualsSql() {
        Expression equalsExpression = RootTable.FIELD_TEST.eq("1");
        assertThat(equalsExpression.buildSql(mDao).first, is("(root.test == ?)"));
    }

    @Test
    public void testNotEqualsSql() {
        Expression notEqualsExpression = RootTable.FIELD_TEST.ne("1");
        assertThat(notEqualsExpression.buildSql(mDao).first, is("(root.test != ?)"));
    }

    @Test
    public void testCount() {
        Expression countExpression = RootTable.FIELD_TEST.count();
        assertThat(countExpression.buildSql(mDao).first, is("COUNT (root.test)"));
    }

    @Test
    public void testCountInHaving() {
        Expression countExpression = RootTable.FIELD_TEST.count().eq(1);
        assertThat(countExpression.buildSql(mDao).first, is("(COUNT (root.test) == 1)"));
    }
}

File: gto-support-db/src/test/java/org/ccci/gto/android/common/db/ExpressionTest.kt
package org.ccci.gto.android.common.db

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.ccci.gto.android.common.db.model.Model1
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

private const val TABLE_NAME = "root"
private const val FIELD_NAME = "test"
private val FIELD = Table.forClass(Model1::class.java).field("test")

@RunWith(AndroidJUnit4::class)
class ExpressionTest {
    private lateinit var dao: AbstractDao

    @Before
    fun setup() {
        dao = mock()
        whenever(dao.tableName(eq(Model1::class.java))).thenReturn(TABLE_NAME)
    }

    @Test
    fun testEqualsSql() {
        assertEquals("($TABLE_NAME.$FIELD_NAME == ?)", FIELD.eq("1").buildSql(dao).first)
    }

    @Test
    fun testNotEqualsSql() {
        assertEquals("($TABLE_NAME.$FIELD_NAME != ?)", FIELD.ne("1").buildSql(dao).first)
    }

    @Test
    fun testCount() {
        assertEquals("COUNT ($TABLE_NAME.$FIELD_NAME)", FIELD.count().buildSql(dao).first)
    }

    @Test
    fun testCountInHaving() {
        assertEquals("(COUNT ($TABLE_NAME.$FIELD_NAME) == 1)", FIELD.count().eq(1).buildSql(dao).first)
    }
}
